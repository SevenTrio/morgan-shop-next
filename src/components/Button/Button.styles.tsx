import styled, { css } from 'styled-components';

import { ISpacingProps } from 'types/spacing.types';

import spacingStyles from 'utils/spacingStyles';

interface IStyledButtonProps extends ISpacingProps {
  variant?: 'primary' | 'secondary';
  fullWidth?: boolean;
}

export const StyledButton = styled.button<IStyledButtonProps>`
  ${spacingStyles};

  display: flex;
  align-items: center;
  justify-content: center;
  width: ${({ fullWidth }) => (fullWidth ? `100%` : `auto`)};
  height: 48px;
  padding: 0 16px;
  border: 1px solid ${({ theme }) => theme.palette.primary.main};
  background-color: transparent;
  transition: background-color 0.2s linear;
  cursor: pointer;

  @media screen and (max-width: ${({ theme }) => theme.breakpoints.sm}px) {
    height: 36px;
    padding: 0 12px;
  }

  &:hover {
    background-color: ${({ theme }) => theme.palette.common.lightGray};
  }

  &:active {
    background-color: ${({ theme }) => theme.palette.common.gray};
  }

  ${({ variant }) =>
    variant === `secondary` &&
    css`
      border: none;
      background-color: ${({ theme }) => theme.palette.primary.main};

      &:hover {
        background-color: ${({ theme }) => theme.palette.primary.light};
      }

      &:active {
        background-color: ${({ theme }) => theme.palette.common.darkGray};
      }
    `}
`;
