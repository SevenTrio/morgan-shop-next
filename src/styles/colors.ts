const colors = {
  black: `#121212`,
  darkGray: `#6D6D6D`,
  gray: `#DEDEDE`,
  lightGray: `#F0F0F0`,
  white: `#FAFAFA`,
  red: `#D92222`,
};

export default colors;
