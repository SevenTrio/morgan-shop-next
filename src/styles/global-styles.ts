import { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
  html {
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  body {
    font-size: 16px;
    font-family: 'Poppins', sans-serif;
  }

  img {
    display: block;
    max-width: 100%;
    overflow: hidden;
  }

  p,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 0;
  }
`;
