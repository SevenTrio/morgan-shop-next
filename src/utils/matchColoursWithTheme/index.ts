import { matchColoursWithTheme } from './matchColoursWithTheme';

export default matchColoursWithTheme;
