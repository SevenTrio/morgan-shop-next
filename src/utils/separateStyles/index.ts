import { separateStyles } from './separateStyles';

export default separateStyles;
