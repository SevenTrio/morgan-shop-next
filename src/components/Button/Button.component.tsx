import { ButtonHTMLAttributes, forwardRef } from 'react';

import { ISpacingProps } from 'types/spacing.types';

import Typography from 'components/Typography';

import { StyledButton } from './Button.styles';

export interface IButtonProps
  extends ISpacingProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * The variant to use.
   */
  variant?: 'primary' | 'secondary';
  /**
   * The component used for the root node.
   */
  component?: 'button' | 'a' | 'span';
  /**
   * If true, the button will take up the full width of its container.
   */
  fullWidth?: boolean;
}

export const Button = forwardRef<HTMLButtonElement, IButtonProps>(
  (
    {
      children,
      component = 'button',
      variant = 'primary',
      fullWidth = false,
      ...restProps
    },
    ref,
  ) => {
    const textColor = variant === `primary` ? `textPrimary` : `white`;

    return (
      <StyledButton
        ref={ref}
        as={component}
        variant={variant}
        fullWidth={fullWidth}
        {...restProps}
      >
        <Typography variant="buttonText" color={textColor}>
          {children}
        </Typography>
      </StyledButton>
    );
  },
);
