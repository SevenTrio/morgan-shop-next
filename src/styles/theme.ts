import { DefaultTheme } from 'styled-components';
import { lighten, darken } from 'polished';

import spacing from 'utils/spacing';
import colors from 'styles/colors';

export const theme: DefaultTheme = {
  breakpoints: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1280,
    xl: 1920,
  },
  palette: {
    common: {
      black: colors.black,
      darkGray: colors.darkGray,
      gray: colors.gray,
      lightGray: colors.lightGray,
      white: colors.white,
      red: colors.red,
    },
    primary: {
      light: `#424242`,
      main: colors.black,
      dark: darken(0.08, colors.black),
      contrastText: colors.white,
    },
    secondary: {
      light: lighten(0.08, colors.white),
      main: colors.white,
      dark: darken(0.08, colors.white),
      contrastText: colors.black,
    },
    error: {
      light: lighten(0.08, colors.red),
      main: colors.red,
      dark: darken(0.08, colors.red),
      contrastText: colors.white,
    },
    text: {
      primary: colors.black,
      secondary: colors.darkGray,
      disabled: colors.darkGray,
      hint: colors.darkGray,
    },
    divider: colors.gray,
    background: {
      default: colors.white,
      productCard: colors.lightGray,
      menu: colors.lightGray,
    },
  },
  shadows: [`1px 1px 10px rgba(0, 0, 0, 0.08)`],
  spacing,
  zIndexes: {
    basket: 100,
  },
};
