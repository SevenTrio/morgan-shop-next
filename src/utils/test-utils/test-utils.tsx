import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { Normalize } from 'styled-normalize';

import { GlobalStyle } from 'styles/global-styles';
import { theme } from 'styles/theme';

const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Normalize />
    <GlobalStyle />
    {children}
  </ThemeProvider>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
