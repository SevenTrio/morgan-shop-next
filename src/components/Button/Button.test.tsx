import { render } from 'utils/test-utils';
import { theme } from 'styles/theme';

import Button from 'components/Button';

const whiteColor = theme.palette.common.white;
const primaryColor = theme.palette.primary.main;
const primaryTextColor = theme.palette.text.primary;

describe(`Button component`, () => {
  describe(`renders correctly`, () => {
    test(`without props`, () => {
      const { container } = render(<Button />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
    });

    test(`with children prop`, () => {
      const buttonText = `Lorem ipsum dolor.`;
      const { container, getByText } = render(<Button>{buttonText}</Button>);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(getByText(buttonText)).not.toBeNull();
    });
  });

  describe(`applies correct styles`, () => {
    test(`by default`, () => {
      const buttonText = `Lorem ipsum dolor.`;
      const { container, getByText } = render(<Button>{buttonText}</Button>);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`width`, `auto`);
      expect(tree).toHaveStyleRule(`height`, `48px`);
      expect(tree).toHaveStyleRule(`padding`, `0 16px`);
      expect(tree).toHaveStyleRule(`background-color`, `transparent`);
      expect(tree).toHaveStyleRule(`border`, `1px solid ${primaryColor}`);
      expect(tree).toHaveStyleRule(`cursor`, `pointer`);
      expect(getByText(buttonText)).toHaveStyleRule(`color`, primaryTextColor);
    });

    test(`with secondary variant`, () => {
      const buttonText = `Lorem ipsum dolor.`;
      const { container, getByText } = render(
        <Button variant="secondary">{buttonText}</Button>,
      );
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`height`, `48px`);
      expect(tree).toHaveStyleRule(`padding`, `0 16px`);
      expect(tree).toHaveStyleRule(`background-color`, primaryColor);
      expect(tree).toHaveStyleRule(`border`, `none`);
      expect(tree).toHaveStyleRule(`cursor`, `pointer`);
      expect(getByText(buttonText)).toHaveStyleRule(`color`, whiteColor);
    });

    test(`with fullWidth prop`, () => {
      const { container } = render(<Button fullWidth />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`width`, `100%`);
    });

    test(`to wrapper component`, () => {
      const { container } = render(
        <Button m={2} style={{ display: `none` }} className="Test" />,
      );
      const tree = container.firstChild as HTMLElement;

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule(`margin`, `8px`);
      expect(tree?.style.display).toBe(`none`);
      expect(tree?.className.split(` `).includes(`Test`)).toBe(true);
    });
  });

  describe(`renders correct tag`, () => {
    test(`by default`, () => {
      const { container } = render(<Button />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.nodeName).toBe(`BUTTON`);
    });

    test(`with component prop`, () => {
      const { container } = render(<Button component="a" />);
      const tree = container.firstChild;

      expect(tree).toMatchSnapshot();
      expect(tree?.nodeName).toBe(`A`);
    });
  });
});
