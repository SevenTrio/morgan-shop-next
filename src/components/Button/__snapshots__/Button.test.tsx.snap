// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component applies correct styles by default 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  >
    Lorem ipsum dolor.
  </p>
</button>
`;

exports[`Button component applies correct styles to wrapper component 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  margin: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0 Test"
  style="display: none;"
>
  <p
    class="c1"
    color="textPrimary"
  />
</button>
`;

exports[`Button component applies correct styles with fullWidth prop 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  />
</button>
`;

exports[`Button component applies correct styles with secondary variant 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #FAFAFA;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
  border: none;
  background-color: #121212;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

.c0:hover {
  background-color: #424242;
}

.c0:active {
  background-color: #6D6D6D;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="white"
  >
    Lorem ipsum dolor.
  </p>
</button>
`;

exports[`Button component renders correct tag by default 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  />
</button>
`;

exports[`Button component renders correct tag with component prop 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<a
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  />
</a>
`;

exports[`Button component renders correctly with children prop 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  >
    Lorem ipsum dolor.
  </p>
</button>
`;

exports[`Button component renders correctly without props 1`] = `
.c1 {
  padding: 0;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #121212;
  text-align: inherit;
  -webkit-letter-spacing: 0.02em;
  -moz-letter-spacing: 0.02em;
  -ms-letter-spacing: 0.02em;
  letter-spacing: 0.02em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding: 0 16px;
  border: 1px solid #121212;
  background-color: transparent;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  cursor: pointer;
}

.c0:hover {
  background-color: #F0F0F0;
}

.c0:active {
  background-color: #DEDEDE;
}

@media screen and (max-width:600px) {
  .c0 {
    height: 36px;
    padding: 0 12px;
  }
}

<button
  class="c0"
>
  <p
    class="c1"
    color="textPrimary"
  />
</button>
`;
